### PR TITLE
OIE Policy API updates

### PR DESCRIPTION
**Pull request by gautamborah-okta**
## Description:
- **What's changed?** 

Adding documentation for new OIE policies: ACCESS_POLICY and PROFILE_ENROLLMENT

As part of policy V1 interoperability two V2 policies App Sign On and Profile Enrollment are made available in V1 format. The policy object documentation is added here.

- **Is this PR related to a Monolith release?** 
This PR is related to OIE release only.

### Resolves:

* [OKTA-388808](https://oktainc.atlassian.net/browse/OKTA-388808)

### CODE REVIEW

File: packages/@okta/vuepress-site/docs/reference/api/policy/index.md [hunk 1]
```diff
@@ -477,6 +477,12 @@ Different Policy types control settings for different operations. All Policy typ
 
 [OAuth Authorization Policy](/docs/reference/api/authorization-servers/#policy-object)
 
+[App Sign On Policy](#app-sign-on-policy) <ApiLifecycle access="ie" /><br>
```
Comment (jakubvul-okta) [inline packages/@okta/vuepress-site/docs/reference/api/policy/index.md:480]:
```suggestion
[App Sign-On Policy](#app-sign-on-policy) <ApiLifecycle access="ie" /><br>
```

Comment (jakubvul-okta) [reply packages/@okta/vuepress-site/docs/reference/api/policy/index.md:480]: Please ensure that this is hyphenated in every instance.

File: packages/@okta/vuepress-site/docs/reference/api/policy/index.md [hunk 2]
```diff
@@ -1619,3 +1625,271 @@ Multiple IdP instances can be defined in a single Policy Action. This allows use
     }
   }
 ```
+
+## App Sign On Policy <ApiLifecycle access="ie" /><br>
+
+> **Note:** This feature is only available as a part of the Okta Identity Engine. Please [contact support](mailto:dev-inquiries@okta.com) for further information.
+
+Application sign-on Policy, which determines the extra levels of authentication (if any) that must be performed before a specific Okta application can be invoked, and are always associated with an application via a Mapping. The Okta Identity Engine always evaluates both Okta Sign On Policy and the Sign On Policy for the application. The resulting user experience will be the union of both policies. App Sign On Policies have the type ACCESS_POLICY.
```
Comment (jakubvul-okta) [inline packages/@okta/vuepress-site/docs/reference/api/policy/index.md:1633]:
```suggestion
The app sign-on policy determines the extra levels of authentication (if any) that must be performed before a specific Okta application can be invoked. It is always associated with an app via a Mapping. The Okta Identity Engine always evaluates both Okta sign-on policy and the sign-on policy for the app. The resulting user experience will be the union of both policies. App sign on policies have the type `ACCESS_POLICY`.
```

File: packages/@okta/vuepress-site/docs/reference/api/policy/index.md (hunk 1 above)
Comment (jakubvul-okta) [reply packages/@okta/vuepress-site/docs/reference/api/policy/index.md:480]: I will change it if I'm making another change nearby, but please do a find+replace to get the other instances that I don't change.

File: packages/@okta/vuepress-site/docs/reference/api/policy/index.md [hunk 3]
```diff
@@ -1619,3 +1625,271 @@ Multiple IdP instances can be defined in a single Policy Action. This allows use
...
+| `factorMode`         | String            | The number of factors required to satisfy this assurance level.                                                         | `1FA`, `2FA`                                                                                      |
+| `constraints`        | Object            | A JSON array containing nested authenticator constraint objects which are organized by Authenticator Class.                                                                           | `POSSESSION`, `KNOWLEDGE`                                                                         |
+| `types`              | Array             | The Authenticator Types which are allowable.                                                                            | `SECURITY_KEY`, `PHONE`, `EMAIL`, `PASSWORD`, `SECURITY_QUESTION`                          |
+| `methods`            | Array             | The Authenticator Methods which are allowable.                                                                          | `PASSWORD`, `SECURITY_QUESTION`, `SMS`, `VOICE`, `EMAIL`, `FIDO2` |
```
Comment (jakubvul-okta) [inline packages/@okta/vuepress-site/docs/reference/api/policy/index.md:1719]:
```suggestion
| `methods`            | Array             | The Authenticator Methods which are permitted.                                                                          | `PASSWORD`, `SECURITY_QUESTION`, `SMS`, `VOICE`, `EMAIL`, `FIDO2` |
```

File: packages/@okta/vuepress-site/docs/reference/api/policy/index.md [hunk 4]
```diff
@@ -1619,3 +1625,271 @@ Multiple IdP instances can be defined in a single Policy Action. This allows use
...
+        ],
+        "hardwareProtection": "REQUIRED"
+      }
+    }
+  ],
+  "reauthenticateIn": "PT4H"
+}
+```
+
+## Profile Enrollment Policy <ApiLifecycle access="ie" /><br>
```
Comment (jakubvul-okta) [inline packages/@okta/vuepress-site/docs/reference/api/policy/index.md:1826]:
```suggestion
## Profile Enrollment Policy 

<ApiLifecycle access="ie" /><br>
```

File: packages/@okta/vuepress-site/docs/reference/api/policy/index.md [hunk 5]
```diff
@@ -1619,3 +1625,271 @@ Multiple IdP instances can be defined in a single Policy Action. This allows use
...
+
+> **Note:** This feature is only available as a part of the Okta Identity Engine. Please [contact support](mailto:dev-inquiries@okta.com) for further information.
+
+Profile Enrollemnt policies specify what profile attributes are required for creating new Users via self-service registration, and also can be used for progressive profiling. The type is specified as PROFILE_ENROLLMENT.
```
Comment (jakubvul-okta) [inline packages/@okta/vuepress-site/docs/reference/api/policy/index.md:1830]:
```suggestion
Profile Enrollemnt policies specify what profile attributes are required for creating new Users via self-service registration, and also can be used for progressive profiling. The type is specified as `PROFILE_ENROLLMENT`.
```

File: packages/@okta/vuepress-site/docs/reference/api/policy/index.md [hunk 6]
```diff
@@ -1619,3 +1625,271 @@ Multiple IdP instances can be defined in a single Policy Action. This allows use
...
+| `profileAttributes.required` | (Optional, default `FALSE`) Indicates if this property is required for enrollment                                                 | Boolean | Required | `FALSE`                                                                                                                                                                                                                        |
+| `profileAttributes` | A list of attributes for which to prompt the user during registration or progressive profiling. Where defined on the User schema, these attributes will be persisted in the user's profile. Non-schema attributes may also be added, which will not be persisted to the user's Profile, but will be included in requests to the Registration Inline Hook. `Only 10 or less profile attributes are supported.`                                                         | Array | Required | N/A                                                                                                                                                                                                                        |
```
Comment (jakubvul-okta) [inline packages/@okta/vuepress-site/docs/reference/api/policy/index.md:1892]:
```suggestion
| `profileAttributes` | A list of attributes for which to prompt the user during registration or progressive profiling. Where defined on the User schema, these attributes will be persisted in the User Profile. Non-schema attributes may also be added, which will not be persisted to the User's Profile, but will be included in requests to the Registration Inline Hook. A maximum of 10 Profile properties is supported.                                                         | Array | Required | N/A                                                                                                                                                                                                                        |
```

File: packages/@okta/vuepress-site/docs/reference/api/policy/index.md [hunk 7]
```diff
@@ -1619,3 +1625,271 @@ Multiple IdP instances can be defined in a single Policy Action. This allows use
...
+| `oktaSignOnSettings`      | TODO                                                                                                                                                                       | TODO                                            | Yes                           | null    |
+
+
+### Verification Method Object
+
+A Verification Method describes the means by which the user must be verified. The only supported type is `ASSURANCE`.
+
+Assurance is the degree of confidence that the end-user logging into an application or service is the same end-user who previously enrolled or logged in to the application or service.
+
+Authenticators can be broadly classified into factors. A factor represents the mechanism by which an end-user owns or controls the authenticator. The three classifications are:
```
Comment (jakubvul-okta) [inline packages/@okta/vuepress-site/docs/reference/api/policy/index.md:1706]:
```suggestion
Authenticators can be broadly classified into factors. A factor represents the mechanism by which an end-user owns or controls the authenticator. The three classifications are:
```

You have to classify into at least two things. Maybe the better question is: what is the difference between a factor and an authenticator?

**Review by jakubvul-okta — CHANGES_REQUESTED** [review]

File: packages/@okta/vuepress-site/docs/reference/api/policy/index.md (hunk 7 above)
Comment (gautamborah-okta) [reply packages/@okta/vuepress-site/docs/reference/api/policy/index.md:1706]: This part is copied from https://github.com/okta/okta-developer-docs/blob/master/packages/%40okta/vuepress-site/docs/reference/api/oie-policy-types/index.md, factors and authenticators are same, authenticators are introduced recently to replace old factors.

Comment (antonioruberto-okta) [reply packages/@okta/vuepress-site/docs/reference/api/policy/index.md:1706]: The term factor is still used in OIE but now `a factor` is the classification/type of an authenticator. Whereas an authenticator is what a user actually uses to verify their identity.

Comment (antonioruberto-okta) [reply packages/@okta/vuepress-site/docs/reference/api/policy/index.md:1706]: cc @thanh-hanguyen-okta @royluo-okta Is the dev doc fine as is or should it explain more about what an authenticator is vs what a factor is? Perhaps we have it defined in some other doc that we could link here?

Comment (jakubvul-okta) [reply packages/@okta/vuepress-site/docs/reference/api/policy/index.md]: I think if we did want to include more of an explanation, it should go in the Authenticators Admin section page.

**Review by petermiller-okta — APPROVED** [review]

File: packages/@okta/vuepress-site/docs/reference/api/policy/index.md [hunk 8]
```diff
@@ -1619,3 +1623,272 @@ Multiple IdP instances can be defined in a single Policy Action. This allows use
...
+| `profileAttributes.required` | (Optional, default `FALSE`) Indicates if this property is required for enrollment                                                 | Boolean | Required | `FALSE`                                                                                                                                                                                                                        |
+| `profileAttributes` | A list of attributes for which to prompt the user during registration or progressive profiling. Where defined on the User schema, these attributes will be persisted in the User Profile. Non-schema attributes may also be added, which will not be persisted to the User's Profile, but will be included in requests to the Registration Inline Hook. A maximum of 10 Profile properties is supported.                                                         | Array | Required | N/A                                                                                                                                                                                                                        |
```
Comment (frankmontanaro-okta) [inline packages/@okta/vuepress-site/docs/reference/api/policy/index.md:1892]: should this also mention:
- also deals with app user schema?
- only allowed for READ-WRITE attributes (hidden and read only are not allowed)
- can also implicitly include any required User schema and AppUser schema?

**Review by frankmontanaro-okta — CHANGES_REQUESTED** [review]
Looks good, just a small suggestion on some clarification

File: packages/@okta/vuepress-site/docs/reference/api/policy/index.md [hunk 9]
```diff
@@ -1619,3 +1623,272 @@ Multiple IdP instances can be defined in a single Policy Action. This allows use
...
+| ---                     | ---                                                                                                                                                                       | ---                                             | ---                           | ---     |
+| `access`                  | `ALLOW` or `DENY`                                                                                                                                                         | `ALLOW` or `DENY`                               | Yes                           | N/A     |
+| `verificationMethod`      | A Verification Method describes the means by which the user must be verified. The only supported type is `ASSURANCE`.                                                       | Object                                          | Yes                           | [Default](#app-sign-on-action-default-example)        |
+
+
+### Verification Method Object
+
+A Verification Method describes the means by which the user must be verified. The only supported type is `ASSURANCE`.
```
Comment (pradeepdewda-okta) [inline packages/@okta/vuepress-site/docs/reference/api/policy/index.md:1700]: change to? `The verification method ensures that user must be verified. Only supported type is `ASSURANCE``

File: packages/@okta/vuepress-site/docs/reference/api/policy/index.md [hunk 10]
```diff
@@ -1619,3 +1623,272 @@ Multiple IdP instances can be defined in a single Policy Action. This allows use
...
+
+### Policy conditions
+
+Policy conditions are not supported, conditions are applied at the rule level for this type of policies.
+
+### Policy Rules conditions
+
+The following conditions may be applied to the rules associated with an app sign on policy:
+
+* [People Condition](#people-condition-object)
```
Comment (pradeepdewda-okta) [inline packages/@okta/vuepress-site/docs/reference/api/policy/index.md:1659]: IDX has to doc `people` and `usertype` conditions

Comment (gautamborah-okta) [reply packages/@okta/vuepress-site/docs/reference/api/policy/index.md:1659]: People condition is already present in the doc. UserType condition will be added later.

**Review by frankmontanaro-okta — APPROVED** [review]
approving to unblock follow up to be created https://okta.slack.com/archives/CCA77QVL2/p1621554230175400?thread_ts=1621547796.167300&cid=CCA77QVL2

File: packages/@okta/vuepress-site/docs/reference/api/policy/index.md [hunk 11]
```diff
@@ -1619,3 +1623,272 @@ Multiple IdP instances can be defined in a single Policy Action. This allows use
...
+
+## App Sign On Policy 
+
+<ApiLifecycle access="ie" /><br>
+
+> **Note:** This feature is only available as a part of the Okta Identity Engine. Please [contact support](mailto:dev-inquiries@okta.com) for further information.
+
+The app sign-on policy determines the extra levels of authentication (if any) that must be performed before a specific Okta application can be invoked. It is always associated with an app via a Mapping. The Okta Identity Engine always evaluates both Okta sign-on policy and the sign-on policy for the app. The resulting user experience will be the union of both policies. App sign on policies have the type `ACCESS_POLICY`.
+
+> **Note:** You can have a maximum of 5000 app sign-on policies in an org.
```
[the remark this answers is not in the document]
Comment (gautamborah-okta) [reply packages/@okta/vuepress-site/docs/reference/api/policy/index.md:1635]: The documentation is for API development. Through the API one could create many app sign on policies.

File: packages/@okta/vuepress-site/docs/reference/api/policy/index.md (hunk 9 above)
Comment (gautamborah-okta) [reply packages/@okta/vuepress-site/docs/reference/api/policy/index.md]: I have made the changes.

**Review by jakubvul-okta — APPROVED** [review]

**Pull request merged**